### PR TITLE
Removed undocumented mid-p correction to p-values in exact test of Hardy-Weinberg equilibrium and updated corresponding tests.

### LINKED ISSUE #7392: Possible bug in ExcessHet p-value calculation?
Waiting on @meganshand  to confirm, but I think the factor of 2 in the denominator of https://github.com/broadinstitute/gatk/blob/796af9fc584929956fda5016d5ab90afd672b822/src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java#L196 should not be there.

I've tested a few cases from Table 1 in Wigginton et al. (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1199378/, upon which this implementation is based) to confirm this. I also separately (and hastily/inefficiently) reimplemented the non-recurrence form of the expression in python:

```
import numpy as np
from scipy.special import gammaln, logsumexp

def factorialln(n):
    return gammaln(1. + n)

def log_prob(n_AB, n_A, N):
    n_B = 2 * N - n_A
    n_AA = (n_A - n_AB) / 2
    n_BB = (n_B - n_AB) / 2
    
    return n_AB * np.log(2) + factorialln(N) + factorialln(n_A) + factorialln(n_B) - \
            (factorialln(n_AA) + factorialln(n_AB) + factorialln(n_BB) + factorialln(2 * N))

def p_value_high(n_AB, n_A, N):
    return np.exp(logsumexp([log_prob(x, n_A, N) for x in range(n_AB, n_A + 2, 2)]))

p_value_high(17, 21, 100) # 0.9304235455950692, agrees with Table 1 in Wigginton et al.
```

The difference is non-negligible. Here's the concordance for data simulated from the null of HWE (for 200 samples x 2000 variants, with a Beta(0.5, 1) alternate-allele frequency spectrum, using pyro):
image

Incidentally, I only realized something might be up when inspecting the distribution of p-values produced by running ExcessHet on this simulated data; there was a weird spike just below and up to 0.5, which I didn't understand. Note that the null p-value distribution isn't expected to be uniform, see https://www.genetics.org/content/180/3/1609.

### PR DESCRIPTION
**Pull request by samuelklee**
Work is split into two commits:

- Removed undocumented mid-p correction to p-values in exact test of Hardy-Weinberg equilibrium and updated corresponding unit tests.
- Updated expected ExcessHet values in integration test resources and added an update toggle to GnarlyGenotyperIntegrationTest.

Various scout cleanups as well.

We now report the same value as ExcHet in bcftools. Note that previous values of 3.0103 (corresponding to mid-p values of 0.5) will now be 0.0000. See discussion below and in linked issue for additional details.

Closes #7392.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:140]: Note that this short circuit for calculating a mid p-value is incorrect, since we are calculating one-sided p-values here. I just went ahead and removed the short circuit.

**Comment by samuelklee** [comment]
Oof, looks like there are now a bunch of broken integration tests that check ExcessHet for whatever reason. So let's definitely decide on whether we want to make the switch to mid p-values before I go through those.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:235]: Consequential test changes start here; most of the other changes in the class are just the result of scout-level cleanup.

**Comment by samuelklee** [comment]
Also just realized there’s *yet another* implementation in htsjdk, HardyWeinbergCalculation, so just a reminder to myself to check against that. Looks like a two-sided p-value of sorts is calculated there, but I think it might be non-canonical in that it doesn’t sum only over het counts with likelihoods less than or equal to that of the observed count—it instead sums over all het counts other than that observe. Need to take a closer look.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:198]: BTW, seems like we should use something from QualityUtils for converting this to Phred scale in the calculateEH method, but will let the reviewer decide. Not sure if we will want to e.g. introduce any caps.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:247]: Another thought: I could put a Java implementation of this slower explicit calculation here in the test code. We could check this against the test cases provided in the paper, and then use it to generate expected results for the additional test cases (rather than hardcoding the results of running the python implementation). Might save future developers a bit of time, although I'm not sure how bulletproof such a Java implementation would be...I'm inclined to trust the scipy/numpy implementations of special functions a little more.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:27]: Most of the following changes are inconsequential cleanup, but note we do change the precision criteria here.

**Comment by samuelklee** [comment]
Also added a little something to prevent output of negative-zero scores (since all of those 3.0103s actually became -0.0000s before this fix), see e.g. https://gatk.broadinstitute.org/hc/en-us/community/posts/360056519272-ExcessHet-value-is-0-0000. Not sure if this is something that should be done at the QualityUtils level either.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/gnarlyGenotyper/GnarlyGenotyperEngine.java
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/gnarlyGenotyper/GnarlyGenotyperEngine.java:228]: Not sure why this unused argument was here, but I took it out.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:91]: Not sure why this unused argument was here, but I took it out.

**Comment by samuelklee** [comment]
Now that I've spent a day surveying the landscape of p-values, I'm thinking what must've happened is that an undocumented mid-p-value correction was added to the one-tailed p-value we calculate. This is probably a fine thing to calculate, but such a correction is not discussed in the Wigginton paper cited in the annotation docs.

So we could keep the original calculation, while further clarifying what exactly we are calculating in the docs (and add references, if appropriate). In this case, we can just keep the cleanup and improved tests as a bonus. (As a further bonus, we won't lose our old friend 3.0103!)

Or we could move to this calculation. I'm not sure if the original intent was to be more/less conservative in retaining sites during hard filtering prior to VQSR. But since the filtering threshold used in Best Practices seems very conservative, I would guess that we wanted to err on the side of not throwing out possible variants, even if they are pretty out of HWE. Which makes the choice of a mid p-value puzzling, since it's strictly smaller and will thus lead to more rejections (I think, if I've got everything the right way around!)

Happy to go either direction. In the end, I have just reaffirmed my dislike for p-values, which I had hitherto thought to be saturated.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:140]: See discussion below about the mid-p correction, but I think it's probably fine to remove the short circuit anyway.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:47]: I noted it over in #7122 post merge, but there are a few tests below that were commented out. Not sure if there is a plan to restore these, but I did not update the corresponding ExcessHet values.

**Comment by samuelklee** [comment]
@ldgauthier @meganshand this is ready for review. Please feel free to un/reassign as necessary. @fleharty @vruano if you have some bandwidth, please take a look.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java
Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:36]: Can you add a test that checks if this is false?  That's what we do for the HaplotypeCaller tests as a failsafe.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (ldgauthier) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:127]: Thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (ldgauthier) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:231]: I'm not sure why this doesn't already implement `ReducibleAnnotation`, but that's why there used to be an `@Override`.  Can you check to see what happens if you add that interface?  It's not unlikely there's a reason this is the exception, but I'd like to remember what it is and document it.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (ldgauthier) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:91]: I may have done a sloppy refactor -- it looks like I'm the one who last changed the method signature.  I'm not worried about this, but thanks for the cleanup.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (ldgauthier) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:36]: There may be a 4.2.2.1 as well...

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java
Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:47]: The tests pass on master if I add them back in. I think I just forgot to put them back.  (I don't have an efficient and elegant method for running the debugger on the Nth test of a data provider if N > 0 so I just commented out 0 through N-1.)  You have my permission to uncomment and update the exact match again.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:28]: We require more precision now, yes?

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:42]: These things are like dandelions -- you take one out and another two pop up.  This is duplicated code that's also in VariantContextTestUtils (and a bunch of other places), but if you feel like you've already done your boy scout duty you can leave it.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:70]: I didn't think about that, so I'm extra glad you're testing it!

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:266]: I don't expect to have to modify this code in the future, so don't break your back bending over backwards for future developers.  I think using a public python library in a separate file is great.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:306]: Nice!

**Review by ldgauthier — CHANGES_REQUESTED** [review]
I'm running the warp plumbing tests now (50 exomes incl. NA12878) to see what changes.  The only possible consequence would be changing the filtering threshold in WARP: https://github.com/broadinstitute/warp/blob/62e06b060b11e246c2623bd68351b2d15e3808bd/pipelines/broad/dna_seq/germline/joint_genotyping/JointGenotyping.wdl#L51.  Is there any theoretical reason we should change it now that mid-p is gone?

**Comment by ldgauthier** [comment]
Now that I think more about it, I'm surprised none of the GenotypeGVCFs tests changed.  I guess they must all be too few samples.

Mostly for my own reference, WARP tests are running here: https://gotc-jenkins.dsp-techops.broadinstitute.org/job/warp-workflow-tests/11738/

**Comment by ldgauthier** [comment]
GenotypeGVCFs results look great (no differences in metrics), but I borked the gnarly docker image name.  Gnarly tests running here: https://gotc-jenkins.dsp-techops.broadinstitute.org/job/warp-workflow-tests/11755/

**Comment by ldgauthier** [comment]
We do see a slight difference in the results, but it's an improvement!(except two indels that are probably in the noise)

## GenotypeGVCFs
### Test:
type	TP	FP	FN	RECALL	PRECISION
SNP	17441	85	846	0.954	0.995
INDEL	304	91	99	0.754	0.77

### Truth:
type	TP	FP	FN	RECALL	PRECISION
SNP	17431	89	856	0.953	0.995
INDEL	305	92	98	0.757	0.768

## Gnarly
Test:
type	TP	FP	FN	RECALL	PRECISION
SNP	17514	137	773	0.958	0.992
INDEL	304	92	99	0.754	0.768

### Truth:
type	TP	FP	FN	RECALL	PRECISION
SNP	17509	152	778	0.957	0.991
INDEL	305	93	98	0.757	0.766

That's good enough for this squirrel.  :shipit:

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHetUnitTest.java:266]: For my part, at least, I have been thinking about parameterizing the annotation to allow the exact test to be performed when an estimate of inbreeding/outcrossing rate is available; it's mentioned near the end of Wigginton et al. as an easy modification to the standard recurrence relation. So I might revisit this code shortly.

But thanks---certainly, the less non-trivial test code I have to write for future developers, the better!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:36]: OK, thanks for the heads up---happy to coordinate with @droazen.

**Comment by samuelklee** [comment]
Thanks for the review and running those tests, @ldgauthier! Will restore the aforementioned tests and update a few other exact matches in the rebase this afternoon.

You also asked above if there was a theoretical reason to change the threshold. Since it seems the original was relatively arbitrary (at least from what I've been told, happy to be corrected), I think we can leave it. The new annotation is strictly larger, so we will then be slightly more conservative about keeping sites if we leave the threshold fixed. You can think of this as a slight change in the decision boundary in genotype-count space---perhaps I can add some plots to this thread this afternoon to demonstrate.

In practice, what we care about is whether: 1) many sites flicker across the change in boundary after hard filtering, and/or 2) these sites result in discrepancies post-VQSR. I think the tests you ran suggest that we don't need to worry much about the second issue, and I can take a closer look later to check about the first (which will depend simply on the number of samples and the allele frequency spectrum). We can also take a basic look at how things might change with e.g. more samples using the aforementioned plots.

**Comment by samuelklee** [comment]
Ah, now that I'm going through and re-updating test resources for a rebase, I see that I missed responding to this:

> Now that I think more about it, I'm surprised none of the GenotypeGVCFs tests changed. I guess they must all be too few samples.

I think some of them did? E.g. I'm having to re-update `src/test/resources/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs/expected/gvcf.basepairResolution.includeNonVariantSites.vcf`.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java:231]: Ah, thanks for that clarification! Hmm, looks like it's missing `getPrimaryRawKey`, `getSecondaryRawKeys`, and `hasSecondaryRawKeys` methods. Not sure what exactly these are supposed to return, but happy to add them if you can enlighten me!

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GnarlyGenotyperIntegrationTest.java:47]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java]: Changed to 4.2.2.0 for now. @droazen let's get this in for 4.2.2.1? Would rather not have to change it again (or deal with any more rebased exact match updates, for that matter!)

**Comment by samuelklee** [comment]
OK, thanks @ldgauthier, I think I've addressed all the comments but one. A little TODO list for my benefit:

- [x] Updated the GATK version in the ExcessHet documentation to 4.2.2.0, but we'll see if I need to revisit that.
- [ ] Not quite sure about the `ReducibleAnnotation` business. Let me know how to make these changes, or else happy to punt and file an issue.
- [ ] I'm not sure I've quite parsed the results of the Jenkins tests, at least in terms of comparing how many sites get hard filtered with/out the change. Where should I be looking at to see the baseline result for that step? Also looks like a lot of results for https://gotc-jenkins.dsp-techops.broadinstitute.org/job/warp-workflow-tests/11755/ were call-cached, is that to be expected? Haven't looked at these tests before, so maybe you can walk me through them at some point. But I guess we can be sure that the overall results don't change too much (at least for 50 samples), which is a good start.
- [ ] Didn't quite get to making those plots of the change in decision boundary, will do that tomorrow or later this week.

**Comment by ldgauthier** [comment]
Based on gs://broad-gotc-test-results/staging/joint_genotyping/exome/scientific/2021-09-03-11-25-15/gather_vcfs_low_memory/small_callset_high_threshold.vcf.gz (from the console output) there are slightly fewer variants filtered with ExcessHet now, which is expected since you said it was an across-the-board shift.  Expected (old) has 4335 and actual (new) has 4133 -- no new things, just some now pass.  If you can calculate a new equivalent threshold I'd rather use that, but otherwise I'm not overly concerned about the changes.

I'm not concerned about the Jenkins call caching unless it's for the GenotypeGVCFs task where ExcessHet actually gets calculated.

For the ReducibleAnnotation comments, if you just revert your changes (statics, visibility, etc.) and open an issue I'm fine with that.  Admittedly this could be another target for refactoring.

**Comment by samuelklee** [comment]
Thanks for those numbers! I'm not sure we can shift the threshold in a uniquely defined way, since things are a function of the allele-frequency spectrum and the number of (non-missing) samples. So definitely glad to hear the impact is minimal and everything behaves as expected.

I'll revert the changes tomorrow and will merge after your final thumbs up, thanks!

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ExcessHet.java]: OK, I've reverted these changes and filed https://github.com/broadinstitute/gatk/issues/7564.

**Comment by samuelklee** [comment]
Seems like there weren't any exact-match tests to update after the latest release, so this should be ready pending your approval, @ldgauthier!

**Review by ldgauthier — APPROVED** [review]
I've held this up long enough, but it might be worth asking @gbrandt6 if she wants to take a look at the ExcessHet documentation changes in that class.

**Comment by samuelklee** [comment]
Thanks @ldgauthier. @gbrandt6 I’d appreciate it if you want to take a look, but I might ask if you can do it by Friday afternoon—I’m out after then through all of next week. Would like to merge before I head out to avoid any more rebasing and/or updating of exact-match tests. Happy to look at any changes to the docs you might make in a subsequent PR, though!

**Comment by samuelklee** [comment]
Actually going to go ahead and merge. Got my booster earlier today and might be knocked out tomorrow. Again, happy to look at subsequent doc changes.

**Comment by gbrandt6** [comment]
@samuelklee I didn't get a chance to take a look, I'll see if we need to make any doc changes next week and let you know

**Pull request merged**
